agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,677 @@ package gateways
+		return err
 	}
 
-	metaGateway.Gateway = apiGateway.(*structs.APIGatewayConfigEntry)
+	meta := ensureInitializedMeta(gateway, bound)
 
-	boundGateway, err := r.store.GetConfigEntry(structs.BoundAPIGateway, req.Name, req.Meta)
+	certificateErrors, err := meta.checkCertificates(store)
 	if err != nil {
-		return nil, err
+		logger.Error("error checking gateway certificates", "error", err)
+		return err
+	}
+
+	for ref, err := range certificateErrors {
+		updater.SetCondition(structs.Condition{
+			Type:               "Accepted",
+			Status:             "False",
+			Reason:             "InvalidCertificate",
+			Message:            err.Error(),
+			Resource:           pointerTo(ref),
+			LastTransitionTime: now,
+		})
+	}
+
+	if len(certificateErrors) > 0 {
+		updater.SetCondition(structs.Condition{
+			Type:               "Accepted",
+			Status:             "False",
+			Reason:             "InvalidCertificates",
+			Message:            "gateway references invalid certificates",
+			LastTransitionTime: now,
+		})
+	} else {
+		updater.SetCondition(structs.Condition{
+			Type:               "Accepted",
+			Status:             "True",
+			Reason:             "Accepted",
+			Message:            "gateway is valid",
+			LastTransitionTime: now,
+		})
+	}
+
+	// now we bind all of the routes we can
+	updatedRoutes := []structs.ControlledConfigEntry{}
+	for _, route := range routes {
+		routeUpdater := structs.NewStatusUpdater(route)
+		_, boundRefs, bindErrors := BindRoutesToGateways([]*gatewayMeta{meta}, route)
+
+		// unset the old gateway binding in case it's stale
+		for _, parent := range route.GetParents() {
+			if parent.Kind == gateway.Kind && parent.Name == gateway.Name && parent.EnterpriseMeta.IsSame(&gateway.EnterpriseMeta) {
+				routeUpdater.RemoveCondition(structs.Condition{
+					Type:     "Bound",
+					Resource: pointerTo(parent),
+				})
+			}
+		}
+
+		// set the status for parents that have bound successfully
+		for _, ref := range boundRefs {
+			routeUpdater.SetCondition(structs.Condition{
+				Type:               "Bound",
+				Status:             "True",
+				Reason:             "Bound",
+				Resource:           pointerTo(ref),
+				Message:            "successfully bound route",
+				LastTransitionTime: now,
+			})
+		}
+
+		// set the status for any parents that have errored trying to
+		// bind
+		for ref, err := range bindErrors {
+			routeUpdater.SetCondition(structs.Condition{
+				Type:               "Bound",
+				Status:             "False",
+				Reason:             "FailedToBind",
+				Resource:           pointerTo(ref),
+				Message:            err.Error(),
+				LastTransitionTime: now,
+			})
+		}
+
+		// if we've updated any statuses, then store them as needing
+		// to be updated
+		if entry, updated := routeUpdater.UpdateEntry(); updated {
+			updatedRoutes = append(updatedRoutes, entry)
+		}
+	}
+
+	// first check for gateway conflicts
+	for i, listener := range meta.BoundGateway.Listeners {
```

Comment:
Can we use `meta.checkConflicts()` here?

Reply:
Unfortunately not in its current form because of the different scope of the status updater. I'm sure we can adjust things slightly in a refactor though. WDYT about me adding a TODO to clean this up in a refactor subsequently?

Reply:
TODO is fine w/ me